{"metadata": {"kernelspec": {"display_name": "Python 3", "language": "python", "name": "python3"}, "language_info": {"name": "python", "version": "3.6.5", "mimetype": "text/x-python", "codemirror_mode": {"name": "ipython", "version": 3}, "pygments_lexer": "ipython3", "nbconvert_exporter": "python", "file_extension": ".py"}}, "nbformat_minor": 4, "nbformat": 4, "cells": [{"cell_type": "code", "source": "chrom = None", "metadata": {"tags": ["parameters"]}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%time\n\nimport pyspark\nimport dxpy\nimport hail as hl\nimport pandas as pd\n\n\nmy_database = dxpy.find_one_data_object(\n    name=\"my_database\", \n    project=dxpy.find_one_project()[\"id\"]\n)[\"id\"]\ndatabase_dir = f'dnax://{my_database}'\nsc = pyspark.SparkContext()\nspark = pyspark.sql.SparkSession(sc)\nhl.init(sc=sc, tmp_dir=f'{database_dir}/tmp/')\n", "metadata": {"trusted": true}, "execution_count": 1, "outputs": [{"name": "stderr", "text": "pip-installed Hail requires additional configuration options in Spark referring\n  to the path to the Hail Python module directory HAIL_DIR,\n  e.g. /path/to/python/site-packages/hail:\n    spark.jars=HAIL_DIR/hail-all-spark.jar\n    spark.driver.extraClassPath=HAIL_DIR/hail-all-spark.jar\n    spark.executor.extraClassPath=./hail-all-spark.jarRunning on Apache Spark version 2.4.4\nSparkUI available at http://ip-10-60-98-96.eu-west-2.compute.internal:8081\n", "output_type": "stream"}, {"name": "stdout", "text": "CPU times: user 3.33 s, sys: 3.66 s, total: 6.99 s\nWall time: 7.78 s\n", "output_type": "stream"}, {"name": "stderr", "text": "Welcome to\n     __  __     <>__\n    / /_/ /__  __/ /\n   / __  / _ `/ / /\n  /_/ /_/\\_,_/_/_/   version 0.2.78-b17627756568\nLOGGING: writing to /opt/notebooks/hail-20220817-0959-0.2.78-b17627756568.log\n", "output_type": "stream"}]}, {"cell_type": "code", "source": "%%time\n\nchrom = chrom\nblock = '*'\n\nvcf_dir='file:///mnt/project/Bulk/Exome sequences_Alternative exome processing/Exome variant call files (gnomAD) (VCFs)'\nvcf_path=f'{vcf_dir}/ukb24068_c{chrom}_b{block}_v1.vcf.gz'\n\n# SPECIFIC TO CHR21\n# assert len(hl.hadoop_ls(vcf_path))==400\n\nraw = hl.import_vcf(\n    vcf_path, \n    force_bgz=True,\n    reference_genome='GRCh38'\n)", "metadata": {"trusted": true}, "execution_count": 2, "outputs": [{"name": "stdout", "text": "CPU times: user 153 ms, sys: 43.3 ms, total: 197 ms\nWall time: 4min 23s\n", "output_type": "stream"}]}, {"cell_type": "code", "source": "def get_ukb_genetic_sex():\n    # NOTE:\n    # 0 = Female\n    # 1 = Male\n    # \"\" = Undefined\n    sex = hl.import_table(\n        'file:///mnt/project/resources/ukb11867_sex.tsv', \n        key='eid', \n        types={\n            'eid': hl.tstr, \n            '22001-0.0_geneticsex': hl.tint\n        },\n        missing=''\n    )\n\n    sex = sex.rename({'eid':'s'})\n    \n    return sex\n\ndef get_is_sex_defined_expr(mt):\n    sex = get_ukb_genetic_sex()\n    return hl.is_defined(sex[mt.s]['22001-0.0_geneticsex'])\n\ndef get_maf_expr(mt):\n    return hl.min(mt.info.AF)\n\ndef get_call_rate_expr(mt):\n    return hl.agg.mean(hl.is_defined(mt.GT))\n\ndef get_mean_depth_expr(mt):\n    return hl.agg.mean(mt.DP)\n\ndef variant_hard_filter_v2_0(mt):\n    # Get high quality, high coverage variants:\n    # \n    # AF > 0.1%\n    # call rate > 0.99\n    # SNV\n    # bi-allelic\n    # does not fail interval qc\n    \n    VARIANT_MAF_MIN = 0.001\n    VARIANT_CALL_RATE_MIN = 0.99\n\n    pass_maf = get_maf_expr(mt) > VARIANT_MAF_MIN\n    pass_call_rate = get_call_rate_expr(mt) > VARIANT_CALL_RATE_MIN\n    pass_is_snv = hl.is_snp(ref=mt.alleles[0], alt=mt.alleles[1])\n    pass_is_biallelic = mt.info.n_alt_alleles==1\n    \n    return mt.filter_rows(\n        pass_maf\n        & pass_call_rate\n        & pass_is_snv\n        & pass_is_biallelic\n        & ~mt.info.fail_interval_qc\n    )\n\ndef export_table(ht, fname, out_folder):\n    ht.naive_coalesce(1).export(f'file:///opt/notebooks/{fname}')\n\n    dxpy.upload_local_file(\n        filename=f'/opt/notebooks/{fname}',\n        name=fname,\n        folder=out_folder,\n        parents=True\n    )\n    \ndef get_unfiltered_row_count_tsv_fname(chrom):\n    return f'variant_count.unfiltered.c{chrom}.tsv'\n\ndef get_hard_filter_row_count_tsv_fname(chrom):\n    return f'variant_count.hq_hard_filter.c{chrom}.tsv'\n    \ndef export_row_count_as_tsv(mt, chrom, fname):\n    row_ct = mt.count_rows()\n    \n    df = pd.DataFrame(data={'row_count': [row_ct]})\n    ht = hl.Table.from_pandas(df)\n        \n    out_folder = '/data/01_calc_call_rate_and_coverage'\n\n    export_table(\n        ht=ht, \n        fname=fname, \n        out_folder=out_folder\n    )", "metadata": {"trusted": true}, "execution_count": 11, "outputs": []}, {"cell_type": "code", "source": "%%time\n\nmt = raw\n\nexport_row_count_as_tsv(mt=mt, chrom=chrom, fname=get_unfiltered_row_count_tsv_fname(chrom))\n\n# Get high quality variants\nmt = variant_hard_filter_v2_0(mt)\n\nexport_row_count_as_tsv(mt=mt, chrom=chrom, fname=get_hard_filter_row_count_tsv_fname(chrom))\n\nmt = mt.select_cols(\n    n_calls = hl.agg.sum(hl.is_defined(mt.GT)),\n    sum_dp = hl.agg.sum(mt.DP),\n)\nht = mt.cols()\n\nfname = f'sample_call_rate_and_coverage_c{chrom}.tsv.gz'\nout_folder = '/data/01_calc_call_rate_and_coverage'\n\nexport_table(\n    ht=ht, \n    fname=fname, \n    out_folder=out_folder\n)", "metadata": {"trusted": true}, "execution_count": null, "outputs": [{"name": "stderr", "text": "2022-08-17 10:35:37 Hail: INFO: Coerced sorted dataset\n", "output_type": "stream"}]}]}