{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "import pyspark\n",
    "import dxpy\n",
    "import hail as hl\n",
    "\n",
    "my_database = dxpy.find_one_data_object(\n",
    "    name=\"my_database\", \n",
    "    project=dxpy.find_one_project()[\"id\"]\n",
    ")[\"id\"]\n",
    "database_dir = f'dnax://{my_database}'\n",
    "sc = pyspark.SparkContext()\n",
    "spark = pyspark.sql.SparkSession(sc)\n",
    "hl.init(sc=sc, tmp_dir=f'{database_dir}/tmp/')\n",
    "\n",
    "chrom = 21\n",
    "block = '*'\n",
    "\n",
    "vcf_dir='file:///mnt/project/Bulk/Exome sequences_Alternative exome processing/Exome variant call files (gnomAD) (VCFs)'\n",
    "vcf_path=f'{vcf_dir}/ukb24068_c{chrom}_b{block}_v1.vcf.gz'\n",
    "\n",
    "# SPECIFIC TO CHR21\n",
    "assert len(hl.hadoop_ls(vcf_path))==400\n",
    "\n",
    "raw = hl.import_vcf(\n",
    "    vcf_path, \n",
    "    force_bgz=True,\n",
    "    reference_genome='GRCh38'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def site_filter(mt):\n",
    "    MIN_DP = 10\n",
    "    MIN_GQ = 20\n",
    "\n",
    "    pass_dp = mt.DP>=MIN_DP\n",
    "    pass_gq = mt.GQ>=MIN_GQ\n",
    "\n",
    "    pass_ab_het = mt.GT.is_het() & (mt.AD[1]/mt.DP>0.2)\n",
    "    pass_ab = ~mt.GT.is_het() | pass_ab_het\n",
    "    mt = mt.filter_entries(pass_dp & pass_gq & pass_ab)\n",
    "\n",
    "    return mt\n",
    "\n",
    "def rf_filter(mt):\n",
    "    mt = mt.filter_rows(mt.filters.contains('RF'), keep=False)\n",
    "    return mt\n",
    "\n",
    "def export_table(ht, fname, out_folder):\n",
    "    ht.naive_coalesce(1).export(f'file:///opt/notebooks/{fname}')\n",
    "\n",
    "    dxpy.upload_local_file(\n",
    "        filename=f'/opt/notebooks/{fname}',\n",
    "        name=sample_qc_fname,\n",
    "        folder=out_folder,\n",
    "        parents=True\n",
    "    )\n",
    "    \n",
    "def export_sample_qc(mt, sample_qc_fname):\n",
    "    mt = hl.sample_qc(mt)\n",
    "\n",
    "    cols = mt.cols()\n",
    "    cols = cols.transmute(**cols.sample_qc.flatten())\n",
    "\n",
    "    sample_qc_path = f'{sample_qc_fname}'\n",
    "\n",
    "    export_table(\n",
    "        ht=cols,\n",
    "        fname=sample_qc_fname,\n",
    "        out_folder='/data/01_get_sample_stats'\n",
    "    )\n",
    "    \n",
    "def export_variant_qc(mt, variant_qc_fname):\n",
    "    mt = hl.variant_qc(mt)\n",
    "\n",
    "    rows = mt.rows()\n",
    "    rows = rows.transmute(**rows.variant_qc.flatten())\n",
    "\n",
    "    variant_qc_path = f'{variant_qc_fname}'\n",
    "\n",
    "    export_table(\n",
    "        ht=rows,\n",
    "        fname=variant_qc_fname,\n",
    "        out_folder='/data/01_get_sample_stats'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%time\n",
    "\n",
    "# raw = site_filter(raw)\n",
    "# raw = rf_filter(raw)\n",
    "\n",
    "# export_sample_qc(\n",
    "#     mt=raw,\n",
    "#     sample_qc_fname=f'sample_qc_c{chrom}.papermill.tsv.gz'\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "\n",
    "raw = site_filter(raw)\n",
    "raw = rf_filter(raw)\n",
    "\n",
    "export_variant_qc(\n",
    "    mt=raw,\n",
    "    variant_qc_fname=f'variant_qc_c{chrom}.tsv.gz',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%time\n",
    "\n",
    "# try:\n",
    "#     raw = raw.annotate_rows(call_rate = hl.agg.mean(hl.is_defined(raw.GT)))\n",
    "\n",
    "#     export_table(\n",
    "#         ht=raw.cols(),\n",
    "#         fname='ukb_wes_450k.rf_gt_filter.variant_call_rate.tsv.gz',\n",
    "#         out_folder='/data/01_get_sample_stats'\n",
    "#     )\n",
    "# except:\n",
    "#     print('Oops, this failed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b''\n"
     ]
    }
   ],
   "source": [
    "# def main():\n",
    "#     import pyspark\n",
    "#     import dxpy\n",
    "#     import hail as hl\n",
    "    \n",
    "#     my_database = dxpy.find_one_data_object(\n",
    "#         name=\"my_database\", \n",
    "#         project=dxpy.find_one_project()[\"id\"]\n",
    "#     )[\"id\"]\n",
    "#     database_dir = f'dnax://{my_database}'\n",
    "#     sc = pyspark.SparkContext()\n",
    "#     spark = pyspark.sql.SparkSession(sc)\n",
    "#     hl.init(sc=sc, tmp_dir=f'{database_dir}/tmp/')\n",
    "    \n",
    "#     chrom = 21\n",
    "#     block = '*'\n",
    "\n",
    "#     vcf_dir='file:///mnt/project/Bulk/Exome sequences_Alternative exome processing/Exome variant call files (gnomAD) (VCFs)'\n",
    "#     vcf_path=f'{vcf_dir}/ukb24068_c{chrom}_b{block}_v1.vcf.gz'\n",
    "\n",
    "#     # SPECIFIC TO CHR21\n",
    "#     assert len(hl.hadoop_ls(vcf_path))==400\n",
    "\n",
    "#     raw = hl.import_vcf(\n",
    "#         vcf_path, \n",
    "#         force_bgz=True,\n",
    "#         reference_genome='GRCh38'\n",
    "#     )\n",
    "    \n",
    "#     def site_filter(mt):\n",
    "#         MIN_DP = 10\n",
    "#         MIN_GQ = 20\n",
    "\n",
    "#         pass_dp = mt.DP>=MIN_DP\n",
    "#         pass_gq = mt.GQ>=MIN_GQ\n",
    "\n",
    "#         pass_ab_het = mt.GT.is_het() & (mt.AD[1]/mt.DP>0.2)\n",
    "#         pass_ab = ~mt.GT.is_het() | pass_ab_het\n",
    "#         mt = mt.filter_entries(pass_dp & pass_gq & pass_ab)\n",
    "\n",
    "#         return mt\n",
    "\n",
    "#     def rf_filter(mt):\n",
    "#         mt = mt.filter_rows(mt.filters.contains('RF'), keep=False)\n",
    "#         return mt\n",
    "\n",
    "#     raw = site_filter(raw)\n",
    "#     raw = rf_filter(raw)\n",
    "    \n",
    "#     def export_sample_qc(mt, sample_qc_fname):\n",
    "#         mt = hl.sample_qc(mt)\n",
    "\n",
    "#         cols = mt.cols()\n",
    "#         cols = cols.transmute(**cols.sample_qc.flatten())\n",
    "\n",
    "#         sample_qc_path = f'file:///opt/notebooks/{sample_qc_fname}'\n",
    "\n",
    "#         cols.naive_coalesce(1).export(sample_qc_path)\n",
    "\n",
    "#         dxpy.upload_local_file(\n",
    "#             filename=f'/opt/notebooks/{sample_qc_fname}',\n",
    "#             name=sample_qc_fname,\n",
    "#             folder='/data/01_get_sample_stats',\n",
    "#             parents=True\n",
    "#         )\n",
    "\n",
    "#     export_sample_qc(\n",
    "#         mt=raw,\n",
    "#         sample_qc_fname=f'sample_qc_c{chrom}.papermill.tsv.gz'\n",
    "#     )\n",
    "\n",
    "# def terminate():\n",
    "#     import subprocess \n",
    "\n",
    "#     bashCommand = 'dx terminate \"$( hostname )\"'\n",
    "#     process = subprocess.Popen(bashCommand.split(), stdout=subprocess.PIPE)\n",
    "#     output, error = process.communicate()\n",
    "    \n",
    "# try:\n",
    "#     main()\n",
    "# except:\n",
    "#     print('Failed')\n",
    "# terminate()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
