{"metadata": {"kernelspec": {"display_name": "Python 3", "language": "python", "name": "python3"}, "language_info": {"name": "python", "version": "3.6.5", "mimetype": "text/x-python", "codemirror_mode": {"name": "ipython", "version": 3}, "pygments_lexer": "ipython3", "nbconvert_exporter": "python", "file_extension": ".py"}}, "nbformat_minor": 4, "nbformat": 4, "cells": [{"cell_type": "code", "source": "chrom = None", "metadata": {"tags": ["parameters"]}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "import pyspark\nimport dxpy\nimport hail as hl\n\nWD='/opt/notebooks'", "metadata": {"trusted": true}, "execution_count": 1, "outputs": []}, {"cell_type": "code", "source": "my_database = dxpy.find_one_data_object(\n    name=\"my_database\", \n    project=dxpy.find_one_project()[\"id\"]\n)[\"id\"]\ndatabase_dir = f'dnax://{my_database}'\nsc = pyspark.SparkContext()\nspark = pyspark.sql.SparkSession(sc)\nhl.init(sc=sc, tmp_dir=f'{database_dir}/tmp/')", "metadata": {"trusted": true}, "execution_count": 2, "outputs": [{"name": "stderr", "text": "pip-installed Hail requires additional configuration options in Spark referring\n  to the path to the Hail Python module directory HAIL_DIR,\n  e.g. /path/to/python/site-packages/hail:\n    spark.jars=HAIL_DIR/hail-all-spark.jar\n    spark.driver.extraClassPath=HAIL_DIR/hail-all-spark.jar\n    spark.executor.extraClassPath=./hail-all-spark.jarRunning on Apache Spark version 2.4.4\nSparkUI available at http://ip-10-60-94-127.eu-west-2.compute.internal:8081\nWelcome to\n     __  __     <>__\n    / /_/ /__  __/ /\n   / __  / _ `/ / /\n  /_/ /_/\\_,_/_/_/   version 0.2.78-b17627756568\nLOGGING: writing to /opt/notebooks/hail-20220818-1344-0.2.78-b17627756568.log\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "## S0. Define functions, load data", "metadata": {}}, {"cell_type": "code", "source": "def import_single_chrom_vcf(chrom):\n    chrom = chrom\n    block = '*'\n    \n    vcf_dir='file:///mnt/project/Bulk/Exome sequences_Alternative exome processing/Exome variant call files (gnomAD) (VCFs)'\n    vcf_path=f'{vcf_dir}/ukb24068_c{chrom}_b{block}_v1.vcf.gz'\n\n    return hl.import_vcf(\n        vcf_path, \n        force_bgz=True,\n        reference_genome='GRCh38'\n    )\n\n\ndef get_mad_threshold_tsv_fname(n_mads, classification):\n    return f'ukb_wes_450k.mad_threshold.nmad_{n_mads}.popclass_{classification}.tsv.gz'\n\n\ndef get_pass_mad_threshold_expr(mt, n_mads='4', classification='strict'):\n    mad_fname = get_mad_threshold_tsv_fname(n_mads=n_mads, classification=classification)\n    mad_path = f'file:///mnt/project/data/03_mad_threshold/{mad_fname}'\n#     mad_path = f'file:///opt/notebooks/{mad_fname}'\n    print(mad_path)\n    mad_ht = hl.import_table(\n        mad_path, \n        types={\n            's': hl.tstr, \n            'pass': hl.tbool\n        },\n        key='s',\n        force=True\n    )\n    \n    return mad_ht[mt.s]['pass']\n\ndef get_fail_interval_qc_expr(mt):\n    return mt.info.fail_interval_qc\n\ndef get_lcr_expr(mt):\n    return mt.info.lcr\n\ndef get_segdup_expr(mt):\n    return mt.info.segdup\n\ndef get_filter_contains_rf_expr(mt):\n    return mt.filters.contains('RF')\n\ndef get_inbreeding_coeff(mt):\n    return mt.info.InbreedingCoeff[0]\n\ndef site_filter(mt):\n    # Set genotype to missing if:\n    # - DP < 10\n    # - GQ < 20\n    # - If heterozygous: Alt allele balance <= 0.2\n    \n    SITE_DP_MIN = 10\n    SITE_GQ_MIN = 20\n\n    pass_dp = mt.DP>=SITE_DP_MIN\n    pass_gq = mt.GQ>=SITE_GQ_MIN\n\n    pass_ab_het = mt.GT.is_het() & (mt.AD[1]/mt.DP>0.2)\n    pass_ab = ~mt.GT.is_het() | pass_ab_het\n    mt = mt.filter_entries(pass_dp & pass_gq & pass_ab)\n\n    return mt\n\n\ndef final_variant_filter(mt):\n    # Remove if:\n    # - FILTER row field contains \"RF\" (random forest true positive probability < {threshold})\n    # - Excess heterozygotes (inbreeding coefficient < -0.3)\n    # - Fails gnomAD interval QC\n    # - In low-complexity region\n    # - segdup is true (segment duplication region?)\n    # - No sample has a high quality genotype\n    \n    MIN_INBREEDING_COEFF = -0.3\n    fails_inbreeding_coeff = get_inbreeding_coeff(mt) < MIN_INBREEDING_COEFF\n    \n    fails_any_hq_genotypes = hl.agg.any(hl.is_defined(mt.GT))\n    \n    return mt.filter_rows(\n        get_filter_contains_rf_expr(mt)\n        | fails_inbreeding_coeff\n        | get_fail_interval_qc_expr(mt)\n        | get_lcr_expr(mt)\n        | get_segdup_expr(mt)\n        | fails_any_hq_genotypes,\n        keep=False\n    )\n\ndef export_table(ht, fname, out_folder):\n    ht.naive_coalesce(1).export(f'file:///opt/notebooks/{fname}')\n\n    dxpy.upload_local_file(\n        filename=f'/opt/notebooks/{fname}',\n        name=fname,\n        folder=out_folder,\n        parents=True\n    )\n    ", "metadata": {"trusted": true}, "execution_count": 28, "outputs": []}, {"cell_type": "code", "source": "%%time\n\nchrom = chrom\n\nraw = import_single_chrom_vcf(chrom)", "metadata": {"trusted": true}, "execution_count": 11, "outputs": [{"name": "stdout", "text": "CPU times: user 54.8 ms, sys: 1.71 ms, total: 56.5 ms\nWall time: 54.9 s\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "## S1. Filter", "metadata": {}}, {"cell_type": "code", "source": "%%time\n\nmt = raw\n\npass_mad_threshold_expr = get_pass_mad_threshold_expr(mt, n_mads='4', classification='strict')\nmt = mt.filter_cols(pass_mad_threshold_expr)\n\nmt = site_filter(mt)\n\n# NOTE: Final variant filter MUST come after site filter in order to remove variants where no individuals have high quality genotypes\nmt = final_variant_filter(mt)", "metadata": {"trusted": true}, "execution_count": 29, "outputs": [{"name": "stdout", "text": "file:///mnt/project/data/03_mad_threshold/ukb_wes_450k.mad_threshold.nmad_4.popclass_strict.tsv.gz\n", "output_type": "stream"}, {"name": "stderr", "text": "2022-08-18 13:54:51 Hail: INFO: Loading 30 fields. Counts by type:\n  str: 29\n  bool: 1\n2022-08-18 13:58:55 Hail: INFO: Coerced sorted dataset\n", "output_type": "stream"}, {"name": "stdout", "text": "count: (190854, 418866)\nCPU times: user 173 ms, sys: 43.6 ms, total: 216 ms\nWall time: 7min 46s\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "## S2. Hail `sample_qc`", "metadata": {}}, {"cell_type": "code", "source": "%%time\n\nmt = hl.sample_qc(mt)\n\ncols = mt.cols()\ncols = cols.transmute(**cols.sample_qc.flatten())\n\nexport_table(\n    ht=cols,\n    fname=f'sample_qc.final_stats.c{chrom}.tsv.gz',\n    out_folder='/data/04_final_filter'\n)", "metadata": {"trusted": true}, "execution_count": null, "outputs": [{"name": "stderr", "text": "2022-08-18 14:28:57 Hail: INFO: Coerced sorted dataset\n", "output_type": "stream"}]}, {"cell_type": "code", "source": "# %%time\n\n# mt = hl.variant_qc(mt)\n\n# rows = mt.rows()\n# rows = rows.transmute(**rows.variant_qc.flatten())\n\n# export_table(\n#     ht=rows,\n#     fname=f'variant_qc.final_stats.c{chrom}.tsv.gz',\n#     out_folder='/data/04_final_filter'\n# )", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}]}